contracts/ERC721A.sol

```diff
@@ -645,10 +645,235 @@ contract ERC721A is IERC721A {
+     */
+    function _batchTransferFrom(
+        address from,
+        address to,
+        uint256[] memory tokenIds,
+        bool approvalCheck
+    ) internal virtual {
+        // Sort `tokenIds` to allow batching consecutive ids into single operations.
+        _sort(tokenIds);
+
+        // Mask `from` and `to` to the lower 160 bits, in case the upper bits somehow aren't clean.
+        from = address(uint160(uint256(uint160(from)) & _BITMASK_ADDRESS));
+        uint256 toMasked = uint256(uint160(to)) & _BITMASK_ADDRESS;
+        if (toMasked == 0) _revert(TransferToZeroAddress.selector);
+
+        uint256 totalTokens = tokenIds.length;
+        uint256 totalTokensLeft;
+        uint256 startTokenId;
+        uint256 nextTokenId;
+        uint256 prevOwnershipPacked;
+        uint256 nextOwnershipPacked;
+        uint256 quantity;
+
+        // If `approvalCheck` is true, check if the caller is approved for all token Ids.
+        // If approved for all, disable next approval checks. Otherwise keep them enabled
+        if (approvalCheck)
+            if (isApprovedForAll(from, _msgSenderERC721A())) approvalCheck = false;
+
+        _beforeTokenBatchTransfers(from, to, tokenIds);
+
+        // Underflow of the sender's balance is temporarily possible if the wrong set of token Ids is passed,
+        // but reverts afterwards when ownership is checked.
+        // The recipient's balance can't realistically overflow.
+        // Counter overflow is incredibly unrealistic as `startTokenId` would have to be 2**256.
+        unchecked {
+            // We can directly increment and decrement the balances.
+            _packedAddressData[from] -= totalTokens;
+            _packedAddressData[to] += totalTokens;
+
+            for (uint256 i; i < totalTokens; ) {
+                startTokenId = tokenIds[i];
+                totalTokensLeft = totalTokens - i;
+
+                // Check ownership of `startTokenId`.
+                prevOwnershipPacked = _packedOwnershipOf(startTokenId);
+                if (address(uint160(prevOwnershipPacked)) != from) _revert(TransferFromIncorrectOwner.selector);
+
+                // Updates startTokenId:
+                // - `address` to the next owner.
+                // - `startTimestamp` to the timestamp of transfering.
+                // - `burned` to `false`.
+                // - `nextInitialized` is left unchanged.
+                _packedOwnerships[startTokenId] = _packOwnershipData(
+                    to,
+                    // TODO: Change to this? Costs +115 gas per [i] loop but ensures correctness of NEXT_INITIALIZED in startTokenId.
+                    // If not consecutive set `nextInitialized` to true, otherwise keep it as is.
+                    // (tokenIds[i + 1] != startTokenId + 1 ? _BITMASK_NEXT_INITIALIZED : prevOwnershipPacked & _BITMASK_NEXT_INITIALIZED) |
+                    (prevOwnershipPacked & _BITMASK_NEXT_INITIALIZED) | _nextExtraData(from, to, prevOwnershipPacked)
```

Comment:
@Vectorized Logic should now be fixed and slightly more optimized. There was an issue with `nextInitialized` that could've resulted in it not being set correctly.

I have just a question. Here (and in line 783) we can actually derive the correct `nextInitialized`, but combining them it would come with performance tradeoffs that scale with the number of tokens transferred (can be very roughly approximated to an increase of 115 gas per token transferred). I'm currently looking into how logic can be refactored to decrease the impact of this. 

But I'm wondering, would it be fine to just leave `nextInitialized` unchanged, like it is now? That would be the most performant outcome, with main diff being that the `nextTokenId` slot could be initialized while the flag of `startTokenId` or `nextTokenId - 1` is not updated.

Reply:
It’s ok if the batch transfers don’t touch  `nextInitialized`. Whatever leads to minimal code.